docs(renovate): enhance renovate documentation (#2211)
AmadeusITGroup/otter: docs/renovate/readme.md

  ]
};
```

### Complex case with private registries

> If some of your dependencies are coming from a private registry, you will have to add the token in the `.npmrc` or `.yarnrc.yml` to be able to fetch the new versions and trigger the install in postUpgradeCommands.

Reviewer: ```suggestion
If some of your dependencies are coming from a private registry, you will have to add the token in the `.npmrc` or `.yarnrc.yml` to be able to fetch the new versions and trigger the install in `allowedPostUpgradeCommands`.
```

Author: I am really talking about post-upgrade commands here, I've updated to clarify: `the post-upgrade commands from the presets`

However, it is not recommended to put the tokens directly in the code of your repository, instead you should use secrets in your CI.

A solution can be to set up the authentication at user level in the CI (`~/.npmrc`).

> This is only working when running `npx renovate` directly (instead of using the GitHub action which relies on docker images)

Reviewer: ```suggestion
This is only working when running `npx renovate` directly (instead of using the GitHub action which relies on docker images).
```

    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
      - uses: actions/setup-node@0a44ba7841725637a19e28fa30b79a866c81b0a6 # v4.0.4
        with:
          node-version: 20
      - name: Prepare global npmrc for renovate
        shell: 'bash'
        run: 'cat .npmrc.ci > ~/.npmrc'
      - name: 'Running Renovate'
        run: |
          git config --global user.email 'bot@renovateapp.com'
          git config --global user.name 'Renovate Bot'
          git config --global core.longpaths true
          cd .github/workflows && npx renovate@~37.420.1
        env:
          RENOVATE_TOKEN: ${{ secrets.RENOVATE_GITHUB_TOKEN }}
          GITHUB_COM_TOKEN: ${{ secrets.RENOVATE_GITHUB_TOKEN }}
          PRIVATE_REGISTRY_READ_TOKEN: ${{ secrets.PRIVATE_REGISTRY_READ_TOKEN }}
```

```js
// .github/workflows/config.js
module.exports = {
  gitAuthor: 'Renovate Bot <bot@renovateapp.com>',
  onboarding: false,
  platform: 'github',
  repositories: [
    // List of repositories to scan
    // ...
  ],
  customEnvVariables: {
    CI: 'true',
    PRIVATE_REGISTRY_READ_TOKEN: process.env.PRIVATE_REGISTRY_READ_TOKEN
  },

>   hostRules: [

Reviewer 2: It could be interesting to mention that the hostRules is to allow the Renovate action while the env is for the postUpgradeTask (if I remember well :D)

Author: I am not even sure that the `hostRules` are still needed now that we set the authentication globally in `~/.npmrc`

Author: after some tests, it turns out that this is very much needed :|
I've updated the doc accordingly

    {
      hostType: 'npm',
      matchHost: '<url of your private registry>',
      username: atob(process.env.PRIVATE_REGISTRY_READ_TOKEN)?.split(':')[0],
      password: atob(process.env.PRIVATE_REGISTRY_READ_TOKEN)?.split(':')[1]
    }
  ],
  allowPostUpgradeCommandTemplating: true,
  allowedPostUpgradeCommands: [
    // Used to execute post-install scripts (like version-harmonize)

>     "^(npm|yarn) install$",

Reviewer 3: not sure about the `$`, you may not want to forbid options

Author: the goal is to limit the allowed commands to only what is really needed for our presets, and currently we do not add any options to the install

>     "^(npm|yarn) install$",

Reviewer 4: shouldn't it be `npm ci` ?

>     "^(npm|yarn) (ci|install)$",

Reviewer 2: Is there a need for an additional cat .npmrc.ci > ~/.npmrc in the post upgrade?

Author: no, the copy is done directly in the workflow (not in the postUpgradeCommands`)

    // Used to execute the migration scripts
    "^(npm|yarn) ng update[a-zA-Z0-9.= {}()#@'\/-]*$",

    // Used to regenerate the SDK
    "^(npm|yarn) run spec:upgrade$",

    // Used to execute the migration scripts of ama-sdk
    "^(npm|yarn) exec schematics @ama-sdk\/schematics:migrate[a-zA-Z0-9.= {}()#'\/-]*$",

    // Used for yarn upgrade when using PnP
    "^yarn dlx @yarnpkg/sdks$"
  ]
};
```

```text
//.npmrc.ci
@my-private-scope:registry=<url of your private registry>
//<url of your private registry>:_auth="${PRIVATE_REGISTRY_READ_TOKEN}"
```

> [!NOTE]

> > The workflow of the bot don't have to be in the same repository as yours, you just need to make sure the token have the right rights

Reviewer: ```suggestion
> The workflow of the bot doesn't have to be in the same repository as yours, you just need to make sure the tokens have the right rights.
```
(or "token has")
